### ISSUE TITLE
Disk Cleanup on Mirrors

### ISSUE THREAD

**Issue by exhuma** [devpi/devpi#760]
Today we've run into a "No space left on device" error on our devpi instance and noticed that the mirror of "pypi" has completely filled up the disk.

This was not really surprising as we kept the available disk-space fairly conservative. By setting it up this way I was "expecting" it to fail at some point.

What was surprising however was the difficulty of cleanup. I came across #542 which seemed scary at first. Yet, I gave it a try to delete a file to be able to do *anything* at all really because most API calls failed due to missing free disk-space. Deleting some files helped some, *but* I ran into an interesting issue: When deleting small packages, the `.sqlite` file grows by more bytes than the size of the deleted files, which leads to the weird behaviour that you *occupy more* disk-space by deleting packages.

Because the `pypi` index on our side is a mirror, I decided it was safe to clean it up as missing packages would be refetched next time they are needed and ran the following sledge-hammer of a command:

```
for line in `devpi list`; do devpi remove -y ${line}; done
```

And to my surprise the free disk space *declined* (because of the increase of space in `.sqlite`).

I resorted to just manually removing some big packages like `pyqt` and `sphinx` which gave me some breathing room.

But apart from nuking the index and recreating it I don't see a good option to reclaim disk-space.

Would it not make sense to allow a optional periodic cleanup of a volatile mirror? Something which removes all files older than a specific age?

Also, I noticed that the mirror always downloads all files from a release, even files for a platform which is not needed (say, macosx on a linux platform). Would it not make sense to download only what's really needed?

Maybe I am just using mirrors for a use-case that was not intended.

**Comment by fschulze** [comment]
There are two parts of data here.

1. The metadata for releases stored in .sqlite. It is updated whenever the list of links for a project is requested and it has changed on pypi. When using ``devpi remove`` on a project it will be refetched as soon as the simple page for it is requested again. Due to the way the data is currently stored in .sqlite, the whole list of links is stored for each change. For projects with many releases this can become quite a bit of data. There is a new storage backend in development which stores this more efficiently.

2. The release files. They are stored on the filesystem in a default devpi-server installation (devpi-postgresql stores them in the db as blobs). These are fetched and stored in the devpi-server instance when the file is requested. They should *not* be fetched by just looking at the links on the simple page. For https://m.devpi.net we had some spiders which downloaded everything, a proper robots.txt stopped that for us.

If you don't need to mirror the release files and only need ``root/pypi`` for pip, then there is a new option which can be set with ``devpi index root/pypi mirror_use_external_urls=true``. When a release is requested it will redirect to the original mirror and *not* fetch and store the file in the devpi-server installation. Already existing files will still be served directly. Those can be removed from the file system directly, as an example for default ``root/pypi``: ``[serverdir]/+files/root/pypi/+f``. We have done this for https://m.devpi.net recently and removed ~60GB of mirrored files. Nowadays pypi is stable enough and pip does a lot more caching of its own, so the default behaviour of devpi-server mirror indexes isn't that useful for most people anymore. Also ``mirror`` was always a misnomer, (``caching``)``proxy`` would have been better.

The only way to reclaim the space used by the metadata is doing an export/import cycle. The ``--hard-links`` option is useful for that to keep the disk space requirement low.

**Comment by exhuma** [comment]
Hey, thanks for the reply. The new external URL option sounds interesting and I will give it a try.

I want to apologize if the tone of my original post may have been a bit aggressive. I had to write the post in a hast after a stressful day. So I appreciate the helpful reply even more.

We have now increased the storage size on our devpi instance considerably but I will definitely give the external URL option a try.
